### PR TITLE
Add node cache

### PR DESCRIPTION
**Pull request by quangkeu95**

### CODE REVIEW

File: node/node_cache.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+	Version string   `json:"jsonrpc,omitempty"`
+	ID      int      `json:"id,omitempty"`
+	Method  string   `json:"method,omitempty"`
+	Params  []string `json:"params,omitempty"`
+	Result  json.RawMessage `json:"result,omitempty"`
+}
+
+type NodeCache struct {
+	client        *http.Client
+	url           string
```
Comment (andrewsource147) [inline node/node_cache.go]: We dont this store this value

File: node/node_cache.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+func NewNodeCache() *NodeCache {
+	nodeEndpoint := os.Getenv("NODE_ENDPOINT")
+
+	nc := &NodeCache{
+		client:        &http.Client{},
+		url:           nodeEndpoint,
+		cacheResponse: make(map[string][]byte),
+		mu:            sync.RWMutex{},
+	}
+	go nc.Run()
```
Comment (andrewsource147) [inline node/node_cache.go]: If this internal method, please use lowercase

File: node/node_cache.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+	return nc
+}
+
+func (nc *NodeCache) Run() {
+	ticker := time.NewTicker(10 * time.Second)
+	for {
+		for _, method := range cacheMethods {
```
Comment (andrewsource147) [inline node/node_cache.go]: The efficiency is create multi go routine for each cacheMethods

File: node/node_cache.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+	if err != nil {
+		log.Print(err)
+		return nil, err
+	}
+
+	// proxyReq.Header = make(http.Header)
+	// for h, val := range req.Header {
+	// 	proxyReq.Header[h] = val
+	// }
+	// log.Print(req.Header)
```
Comment (andrewsource147) [inline node/node_cache.go]: Remove comment if this unneeeded

File: node/node_cache.go [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+		if err != nil {
+			log.Print(err)
+			return nil, err
+		}
+		return bodyBytes, nil
+	}
+	return nil, errors.New(fmt.Sprintf("Status code is %s", resp.StatusCode))
+}
+
+func (nc *NodeCache) makeRequest(method string) (*http.Request, error) {
```
Comment (andrewsource147) [inline node/node_cache.go]: Do we need to return a pointer, pls remember blogpost I share you before

File: node/node_cache.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+	return req, nil
+}
+
+func (nc *NodeCache) setCacheResponse(method string, resp []byte) {
+	nc.mu.Lock()
+	defer nc.mu.Unlock()
+	nc.cacheResponse[method] = resp
+}
+
+func (nc *NodeCache) getCacheResponse(method string) []byte {
```
Comment (andrewsource147) [inline node/node_cache.go]: This should be public func

File: node/node.go [hunk 7]
```diff
@@ -115,11 +65,11 @@ func filterRequest(req *http.Request) error {
 	}
 
 	// check origin
-	origin := req.Header.Get("Origin")
-
-	if !InListSubstring(origin, whiteListArr) {
-		return errors.New("Domain is not in the whitelist: " + origin)
-	}
+	//origin := req.Header.Get("Origin")
```
Comment (andrewsource147) [inline node/node.go]: Why you remove check origin?

File: node/node_cache.go (hunk 1 above)
Comment (quangkeu95) [reply node/node_cache.go]: removed

File: node/node_cache.go (hunk 2 above)
Comment (quangkeu95) [reply node/node_cache.go]: converted to lowercase

File: node/node.go (hunk 7 above)
Comment (quangkeu95) [reply node/node.go]: my mistake, uncommented

File: node/node_cache.go (hunk 5 above)
Comment (quangkeu95) [reply node/node_cache.go]: I think we can return pointer, it improves code readability, and http.Request already contained pointer inside.

File: node/node.go [hunk 8]
```diff
@@ -25,8 +26,9 @@ var banMethod = []string{}
 func NewNodeMiddleware() (*NodeMiddleware, error) {
 	nodeEnpoint := os.Getenv("NODE_ENDPOINT")
 	return &NodeMiddleware{
-		client: &http.Client{},
-		url:    nodeEnpoint,
+		client:    &http.Client{},
+		url:       nodeEnpoint,
```
Comment (andrewsource147) [inline node/node.go]: Do we need url here?

Comment (quangkeu95) [reply node/node.go]: removed

**Pull request merged**